src/HappyConnOpener.cc

```diff
@@ -0,0 +1,406 @@
+    HappyConnOpener::LastAttempt = current_dtime;
+    AsyncJob::Start(cs);
+}
+
+void
+HappyConnOpener::connectDone(const CommConnectCbParams &params)
+{
+    if (master.path == params.conn) {
+        // master connection is now the remaining spare if exist, or null
+        master = spare;
+    } else {
+        assert(spare.path == params.conn);
+    }
+
+    if (spare.path != nullptr) {
+        HappyConnOpener::SpareConnects--;
+        spare.path = nullptr;
+        spare.connector = nullptr;
+    }
+
+    if (params.flag != Comm::OK) {
+        debugs(17, 8, "Connections to " << params.conn << " failed");
+        /* it might have been a timeout with a partially open link */
+        if (params.conn != NULL) {
+            if (params.conn->getPeer())
+                peerConnectFailed(params.conn->getPeer());
+
+            params.conn->close();
+        }
+
+        checkForNewConnection();
+        return;
+    }
+
+    debugs(17, 8, "Connections to " << params.conn << " succeed");
+    if (master.path) {
+        Must(master.connector != nullptr);
+        master.connector->cancel("Already connected");
+        master.connector = nullptr;
+        master.path = nullptr;
+    }
+
+    callCallback(params.conn, Comm::OK, 0, false, "new connection");
+    return;
+}
+
+Comm::ConnectionPointer
+HappyConnOpener::getCandidatePath()
+{
+    if (dests_->empty())
+        return Comm::ConnectionPointer();
+
+    // if there is not any pending connection just get the first available destination
+    if (master.path == nullptr)
+        return dests_->popFirst();
+
+    // Check if there is available destination with different protocol
+    // than the last connector
+    int lastConnectionFamily = CandidatePaths::ConnectionFamily(master.path);
+    return dests_->popFirstNotInFamily(lastConnectionFamily);
+}
+
+bool
+HappyConnOpener::existCandidatePath()
+{
+    if (master.path == nullptr)
+        return !dests_->empty();
+
+    int lastConnectionFamily = CandidatePaths::ConnectionFamily(master.path);
+    return dests_->existPathNotInFamily(lastConnectionFamily);
+}
+
+void
+HappyConnOpener::checkForNewConnection()
+{
+    debugs(17, 8, "Check for starting new connection");
+    assert(dests_ != nullptr);
+
+    // If it is already in WaitingConnectors queue wait the CheckForConnectionAttempts()
+    // be executed and start new connection if required.
+    // Starting new connection here will alter the this->nextAttemptTime member
+    // which affects the WaitingConnectors ordering.
+    if (queueSubscribed_) {
+        debugs(17, 8, "already subscribed for starting new connection when ready");
+        return;
+    }
+
+    if (master.path == nullptr || spareConnectionNeeded()) {
+        Comm::ConnectionPointer dest = getCandidatePath();
+        if (dest != nullptr)
+            startConnecting(dest);
+    }
+
+    if (Config.happyEyeballs.connect_limit == 0)
```

Comment:
The checkForNewConnection method check if should start a new connection, a master connection or spare connection.
If spare connections are disabled then should start only master connections and do nothing if waiting for master connection to be established (or rejected).

The check is needed here to avoid schedule a spare connection if the feature is disabled.

I agree the code is not good.

Reply:
I did some changes on this method. 
I think it is easier to understand what it does not, I hope it is OK.